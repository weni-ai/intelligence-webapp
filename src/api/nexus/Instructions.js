import nexusRequest from '../nexusaiRequest';
import { InstructionAdapter } from '../adapters/instructions/instruction';
import { useInstructionsStore } from '@/store/Instructions';

const request = nexusRequest;

export const Instructions = {
  async addInstruction({ projectUuid, instruction }) {
    const instructionsStore = useInstructionsStore();
    const body = {
      instructions: [...instructionsStore.instructions.data, instruction].map(
        InstructionAdapter.toApi,
      ),
    };
    const response = await request.$http.put(
      `api/${projectUuid}/customization/`,
      body,
      {
        hideGenericErrorAlert: true,
      },
    );

    const biggestId = response.data.instructions.reduce((acc, instruction) => {
      if (instruction.id > acc) acc = instruction.id;
      return acc;
    }, 0);

    return InstructionAdapter.fromApi(
      response.data.instructions.find(
        (instruction) => instruction.id === biggestId,
      ),
    );
  },

  async list({ projectUuid }) {
    const response = await request.$http.get(
      `api/${projectUuid}/customization/`,
    );

    return response.data.instructions.map(InstructionAdapter.fromApi);
  },

  async edit({ projectUuid, id, text }) {
    const instructionsStore = useInstructionsStore();

    const body = {
      instructions: [
        ...instructionsStore.instructions.data.filter(
          (instruction) => instruction.id !== id,
        ),
        { id, text },
      ].map(InstructionAdapter.toApi),
    };
    await request.$http.put(`api/${projectUuid}/customization/`, body);
  },

  async delete({ projectUuid, id }) {
    await request.$http.delete(`api/${projectUuid}/customization/?id=${id}`);
  },
};
