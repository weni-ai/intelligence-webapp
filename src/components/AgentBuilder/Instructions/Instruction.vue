<template>
  <section class="instruction">
    <p
      class="instruction__text"
      data-testid="instruction-text"
    >
      {{ text }}
    </p>

    <p
      v-if="tag"
      class="instruction__tag"
      data-testid="instruction-tag"
    >
      {{ tag }}
    </p>
  </section>
</template>

<script setup>
const props = defineProps({
  text: {
    type: String,
    required: true,
  },
  tag: {
    type: String,
    default: '',
    required: false,
  },
});
</script>

<style lang="scss" scoped>
.instruction {
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: $unnnic-spacing-sm;

  padding: $unnnic-spacing-sm;

  &:not(:last-child) {
    border-bottom: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;
  }

  &__text {
    margin: 0;
    color: $unnnic-color-neutral-cloudy;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__tag {
    border-radius: $unnnic-border-radius-pill;

    padding: $unnnic-spacing-nano $unnnic-spacing-ant;

    background-color: $unnnic-color-neutral-light;
    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-md;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
  }
}
</style>
