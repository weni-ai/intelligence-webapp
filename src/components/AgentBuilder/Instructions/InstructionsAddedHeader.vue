<template>
  <header class="instructions-added-header">
    <hgroup
      class="instructions-added-header__infos"
      data-testid="instructions-header-infos"
    >
      <h2
        class="infos__title"
        data-testid="instructions-title"
      >
        {{ $t('agent_builder.instructions.instructions_added.title') }}
      </h2>
      <h3
        class="infos__subtitle"
        data-testid="instructions-subtitle"
      >
        {{
          $t('agent_builder.instructions.instructions_added.subtitle', {
            count: instructionsCount,
          })
        }}
      </h3>
    </hgroup>

    <UnnnicSelectSmart
      v-model="filters"
      :options="filtersOptions"
      orderedByIndex
      multiple
      multipleWithoutSelectsMessage
      autocomplete
      data-testid="instructions-filters"
    />
  </header>
</template>

<script setup>
import { computed } from 'vue';
import i18n from '@/utils/plugins/i18n';

defineProps({
  instructionsCount: {
    type: Number,
    required: true,
  },
});

const filters = defineModel('filters', {
  type: Array,
  required: true,
});

const getFilterTranslation = (filter) =>
  i18n.global.t(
    `agent_builder.instructions.instructions_added.filters.${filter}`,
  );
const filtersOptions = computed(() => [
  {
    label: getFilterTranslation('placeholder'),
    value: '',
  },
  {
    label: getFilterTranslation('default_instructions'),
    value: 'default_instructions',
  },
  {
    label: getFilterTranslation('personalized_instructions'),
    value: 'personalized_instructions',
  },
]);
</script>

<style lang="scss" scoped>
.instructions-added-header {
  display: grid;
  grid-template-columns: 9fr 3fr;
  align-items: center;
  gap: $unnnic-spacing-sm;

  &__infos {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-nano;

    .infos__title {
      margin: 0;

      color: $unnnic-color-neutral-darkest;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-bold;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }

    .infos__subtitle {
      margin: 0;

      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-md;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }
  }
}
</style>
