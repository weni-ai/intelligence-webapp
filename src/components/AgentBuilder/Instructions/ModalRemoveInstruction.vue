<template>
  <UnnnicModalDialog
    :modelValue="modelValue"
    data-testid="modal"
    class="modal-remove-instruction"
    showCloseIcon
    icon="warning"
    iconScheme="aux-red-500"
    :title="$t('agent_builder.instructions.remove_instruction.title')"
    size="sm"
    :secondaryButtonProps="{
      text: $t('agent_builder.instructions.remove_instruction.cancel'),
    }"
    :primaryButtonProps="{
      text: $t('agent_builder.instructions.remove_instruction.remove'),
      loading: instruction.status === 'loading',
      type: 'warning',
    }"
    @secondary-button-click="close"
    @primary-button-click="removeInstruction"
    @update:model-value="close"
  >
    <p
      class="modal-remove-instruction__description"
      data-testid="description"
    >
      {{
        $t('agent_builder.instructions.remove_instruction.modal_description')
      }}
    </p>
  </UnnnicModalDialog>
</template>

<script setup>
import { useInstructionsStore } from '@/store/Instructions';
import { ref } from 'vue';

const instructionsStore = useInstructionsStore();

const emit = defineEmits(['update:modelValue']);

const props = defineProps({
  instruction: {
    type: Object,
    required: true,
  },
});

const modelValue = ref(false);

function close() {
  emit('update:modelValue', false);
}

async function removeInstruction() {
  const { status } = await instructionsStore.removeInstruction(
    props.instruction.id,
  );

  if (status === 'complete') close();
}
</script>

<style lang="scss" scoped>
.modal-remove-instruction {
  :deep(.unnnic-modal-dialog__container__content) {
    padding-bottom: $unnnic-spacing-xs;
  }

  &__description {
    color: $unnnic-color-neutral-cloudy;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }
}
</style>
