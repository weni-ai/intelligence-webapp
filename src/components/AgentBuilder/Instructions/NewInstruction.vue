<template>
  <section
    class="new-instruction"
    data-testid="new-instruction"
  >
    <h2
      class="new-instruction__title"
      data-testid="new-instruction-title"
    >
      {{ $t('agent_builder.instructions.new_instruction.title') }}
    </h2>
    <UnnnicTextArea
      v-model="instructionsStore.newInstruction.text"
      data-testid="new-instruction-textarea"
      :placeholder="
        $t('agent_builder.instructions.new_instruction.textarea.placeholder')
      "
      :message="
        $t('agent_builder.instructions.new_instruction.textarea.description')
      "
    />
    <UnnnicButton
      class="new-instruction__add-instruction-button"
      data-testid="add-instruction-button"
      :disabled="!newInstruction.text.trim()"
      :text="$t('agent_builder.instructions.new_instruction.add_instruction')"
      :loading="newInstruction.status === 'loading'"
      @click="instructionsStore.addInstruction"
    />
  </section>
</template>

<script setup>
import { computed } from 'vue';
import { useInstructionsStore } from '@/store/Instructions';

const instructionsStore = useInstructionsStore();
const newInstruction = computed(() => instructionsStore.newInstruction);
</script>

<style lang="scss" scoped>
.new-instruction {
  border-radius: $unnnic-border-radius-md;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;
  padding: $unnnic-spacing-sm;

  display: grid;
  grid-template-columns: repeat(12, 1fr);
  row-gap: $unnnic-spacing-sm;

  & > * {
    grid-column: 1 / -1;
  }

  &__title {
    margin: 0;

    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__add-instruction-button {
    grid-column: 1 / 5;
  }
}
</style>
