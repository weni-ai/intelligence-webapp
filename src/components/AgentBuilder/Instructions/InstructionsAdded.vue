<template>
  <section class="instructions-added">
    <h2
      class="instructions-added__title"
      data-testid="instructions-title"
    >
      {{ $t('agent_builder.instructions.instructions_added.title') }}
    </h2>

    <section
      :class="[
        'instructions-added__instructions',
        {
          'instructions-added__instructions--loading':
            instructionsStore.instructions.status === 'loading',
        },
      ]"
      data-testid="instructions"
    >
      <template v-if="instructionsStore.instructions.status === 'loading'">
        <UnnnicSkeletonLoading
          v-for="index in 10"
          :key="index"
          tag="div"
          width="100%"
          height="60px"
          data-testid="instruction-loading"
        />
      </template>
      <template v-else>
        <Instruction
          v-for="(instruction, index) in instructionsDefault"
          :key="index"
          data-testid="instruction-default"
          :text="instruction"
          :tag="
            $t(
              'agent_builder.instructions.instructions_added.default_instruction',
            )
          "
        />
        <Instruction
          v-for="instruction in instructionsStore.instructions.data"
          :key="instruction.id"
          data-testid="instruction-added"
          :text="instruction.text"
        />
      </template>
    </section>
  </section>
</template>

<script setup>
import { computed } from 'vue';
import { useInstructionsStore } from '@/store/Instructions';

import Instruction from './Instruction.vue';
import i18n from '@/utils/plugins/i18n';

const instructionsStore = useInstructionsStore();
if (instructionsStore.instructions.status === null) {
  instructionsStore.loadInstructions();
}

const instructionsDefault = computed(() =>
  i18n.global.tm(
    'agent_builder.instructions.instructions_added.default_instructions',
  ),
);
</script>

<style lang="scss" scoped>
.instructions-added {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-sm;

  &__title {
    margin: 0;

    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__instructions {
    display: flex;
    flex-direction: column;

    border-radius: $unnnic-border-radius-md;
    border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

    &--loading {
      gap: $unnnic-spacing-nano;

      border: none;
      border-radius: 0;
    }
  }
}
</style>
