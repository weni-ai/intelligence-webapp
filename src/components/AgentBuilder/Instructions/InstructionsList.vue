<template>
  <section
    :class="[
      'instructions-list',
      {
        'instructions-list--loading': isLoading,
        'instructions-list--no-instructions': instructions.length === 0,
      },
    ]"
    data-testid="instructions"
  >
    <template v-if="isLoading">
      <UnnnicSkeletonLoading
        v-for="index in 10"
        :key="index"
        tag="div"
        width="100%"
        height="60px"
        data-testid="instruction-loading"
      />
    </template>
    <section
      v-else-if="instructions.length === 0"
      class="instructions-list__no-instructions"
    >
      <UnnnicIcon
        icon="format_list_bulleted"
        size="xl"
        scheme="neutral-cleanest"
      />
      <p class="no-instructions__text">
        {{
          $t(
            'agent_builder.instructions.instructions_list.no_custom_instructions',
          )
        }}
      </p>
    </section>
    <template v-else>
      <Instruction
        v-for="instruction in instructions"
        :key="instruction.id"
        :instruction="instruction"
        :showActions="showActions"
        data-testid="instruction-added"
      />
    </template>
  </section>
</template>

<script setup>
import Instruction from './Instruction.vue';

defineProps({
  instructions: {
    type: Array,
    required: true,
  },
  isLoading: {
    type: Boolean,
    default: false,
  },
  showActions: {
    type: Boolean,
    default: false,
  },
});
</script>

<style lang="scss" scoped>
.instructions-list {
  margin-bottom: $unnnic-spacing-sm;

  display: flex;
  flex-direction: column;

  border-radius: $unnnic-border-radius-md;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

  &--loading,
  &--no-instructions {
    height: 100%;

    gap: $unnnic-spacing-nano;

    border: none;
    border-radius: 0;
  }

  &__no-instructions {
    height: 100%;

    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: $unnnic-spacing-nano;

    .no-instructions__text {
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-gt;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }
  }
}
</style>
