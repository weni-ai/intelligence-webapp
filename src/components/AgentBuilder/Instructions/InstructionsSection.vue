<template>
  <section class="instructions-section">
    <h2
      class="instructions-section__title"
      data-testid="instructions-section-title"
    >
      {{ $t('agent_builder.instructions.instructions_list.title') }}
    </h2>

    <UnnnicTab
      :tabs="tabs"
      :activeTab="activeTab"
      data-testid="instructions-section-tab"
      @change="instructionsStore.activeInstructionsListTab = $event"
    >
      <template
        v-for="tab in tabs"
        :key="tab"
        #[`tab-head-${tab}`]
      >
        {{ $t(`agent_builder.instructions.instructions_list.tabs.${tab}`) }}
      </template>
      <template #[`tab-panel-${activeTab}`]>
        <p class="instructions-section__tab-description">
          {{
            $t(
              `agent_builder.instructions.instructions_list.tabs.${activeTab}_description`,
            )
          }}
        </p>
      </template>
    </UnnnicTab>
    <InstructionsList
      v-if="activeTab === 'custom'"
      data-testid="instructions-custom"
      :instructions="instructionsStore.instructions.data"
      :isLoading="instructionsStore.instructions.status === 'loading'"
      showActions
    />
    <InstructionsList
      v-if="activeTab === 'default'"
      data-testid="instructions-default"
      :instructions="instructionsDefault"
      :isLoading="false"
      :showActions="false"
    />
    <InstructionsList
      v-if="activeTab === 'safety_topics'"
      data-testid="instructions-safety-topics"
      :instructions="instructionsSafetyTopics"
      :isLoading="false"
      :showActions="false"
    />
  </section>
</template>

<script setup>
import { computed, ref } from 'vue';
import { useInstructionsStore } from '@/store/Instructions';

import InstructionsList from './InstructionsList.vue';
import i18n from '@/utils/plugins/i18n';

const instructionsStore = useInstructionsStore();

if (instructionsStore.instructions.status === null) {
  instructionsStore.loadInstructions();
}

const tabs = ref(['custom', 'default', 'safety_topics']);
const activeTab = computed(() => instructionsStore.activeInstructionsListTab);

function generateMockedInstructions(type) {
  return i18n.global
    .tm(`agent_builder.instructions.instructions_list.${type}`)
    .map((instruction, index) => ({
      id: `${type}-${index}`,
      text: instruction,
    }));
}

const instructionsDefault = computed(() =>
  generateMockedInstructions('default_instructions'),
);

const instructionsSafetyTopics = computed(() =>
  generateMockedInstructions('safety_topics'),
);
</script>

<style lang="scss" scoped>
.instructions-section {
  display: flex;
  flex-direction: column;

  &__title {
    margin: 0 0 $unnnic-spacing-sm;

    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__tab-description {
    margin-bottom: $unnnic-spacing-xs;

    color: $unnnic-color-neutral-cloudy;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }
}
</style>
