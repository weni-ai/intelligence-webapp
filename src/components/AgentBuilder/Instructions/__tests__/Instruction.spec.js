import { shallowMount } from '@vue/test-utils';
import { describe, it, expect, beforeEach } from 'vitest';
import { createTestingPinia } from '@pinia/testing';

import Instruction from '../Instruction.vue';
import i18n from '@/utils/plugins/i18n';
import { useInstructionsStore } from '@/store/Instructions';

describe('Instruction.vue', () => {
  let wrapper;
  let instructionsStore;
  const pinia = createTestingPinia({
    initialState: {
      Instructions: {
        instructions: {
          data: [],
        },
      },
    },
  });

  beforeEach(() => {
    instructionsStore = useInstructionsStore(pinia);

    wrapper = shallowMount(Instruction, {
      global: {
        plugins: [pinia],
      },
      props: {
        instruction: {
          id: 'test-id',
          text: 'Test instruction',
        },
        tag: '',
        showActions: true,
      },
    });
  });

  const SELECTORS = {
    text: '[data-testid="instruction-text"]',
    tag: '[data-testid="instruction-tag"]',
    actions: '[data-testid="instruction-actions"]',
    input: '[data-testid="instruction-input"]',
    saveButton: '[data-testid="instruction-save-button"]',
    cancelButton: '[data-testid="instruction-cancel-button"]',
    modalRemoveInstruction: '[data-testid="modal-remove-instruction"]',
  };
  const find = (selector) => wrapper.find(SELECTORS[selector]);
  const findComponent = (component) =>
    wrapper.findComponent(SELECTORS[component]);

  describe('Component rendering', () => {
    it('should match snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders components correctly', () => {
      expect(find('text').exists()).toBe(true);
      expect(findComponent('actions').exists()).toBe(true);
      expect(findComponent('modalRemoveInstruction').exists()).toBe(true);
      expect(find('tag').exists()).toBe(false);
      expect(find('input').exists()).toBe(false);
      expect(find('saveButton').exists()).toBe(false);
      expect(find('cancelButton').exists()).toBe(false);
    });
  });

  describe('Component props', () => {
    it('renders the correct text', () => {
      expect(find('text').text()).toBe('Test instruction');
    });

    it('renders the correct tag', async () => {
      await wrapper.setProps({ tag: 'Test tag' });
      expect(find('tag').exists()).toBe(true);
      expect(find('tag').text()).toBe('Test tag');
    });
  });

  describe('Content item actions', () => {
    it('renders the component when showActions is true', async () => {
      await wrapper.setProps({ showActions: true });
      expect(findComponent('actions').exists()).toBe(true);
    });

    it('does not render the component when showActions is false', async () => {
      await wrapper.setProps({ showActions: false });
      expect(findComponent('actions').exists()).toBe(false);
    });

    it('passes the correct props', async () => {
      await wrapper.setProps({ showActions: true });
      expect(findComponent('actions').props('actions')).toEqual([
        {
          text: i18n.global.t(
            'agent_builder.instructions.edit_instruction.title',
          ),
          icon: 'edit_square',
          scheme: 'neutral-dark',
          onClick: expect.any(Function),
        },
        {
          text: i18n.global.t(
            'agent_builder.instructions.remove_instruction.title',
          ),
          icon: 'delete',
          scheme: 'aux-red-500',
          onClick: expect.any(Function),
        },
      ]);
    });

    it('closes the modal when the update:model-value event is emitted', async () => {
      wrapper.vm.showModalRemoveInstruction = true;
      expect(wrapper.vm.showModalRemoveInstruction).toBe(true);

      findComponent('modalRemoveInstruction').vm.$emit(
        'update:model-value',
        false,
      );
      expect(wrapper.vm.showModalRemoveInstruction).toBe(false);
    });
  });

  describe('Edit mode', () => {
    beforeEach(() => {
      wrapper.vm.isEditing = true;
      instructionsStore.editInstruction = vi.fn().mockResolvedValue({
        status: 'complete',
      });
    });

    it('should match snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot('edit-mode');
    });

    it('renders the correct components', async () => {
      expect(findComponent('input').exists()).toBe(true);
      expect(findComponent('saveButton').exists()).toBe(true);
      expect(findComponent('cancelButton').exists()).toBe(true);
      expect(findComponent('text').exists()).toBe(false);
      expect(findComponent('tag').exists()).toBe(false);
      expect(findComponent('actions').exists()).toBe(false);
    });

    it('renders the correct input', () => {
      expect(findComponent('input').props('modelValue')).toBe(
        'Test instruction',
      );
      expect(findComponent('input').props('size')).toBe('sm');
    });

    it('renders the correct save button', () => {
      expect(findComponent('saveButton').props('disabled')).toBe(false);
      expect(findComponent('saveButton').props('loading')).toBe(false);
      expect(findComponent('saveButton').props('text')).toBe(
        i18n.global.t('agent_builder.instructions.edit_instruction.save'),
      );
      expect(findComponent('saveButton').props('type')).toBe('secondary');
      expect(findComponent('saveButton').props('size')).toBe('small');
    });

    it('renders the correct cancel button', () => {
      expect(findComponent('cancelButton').props('text')).toBe(
        i18n.global.t('agent_builder.instructions.edit_instruction.cancel'),
      );
      expect(findComponent('cancelButton').props('type')).toBe('tertiary');
      expect(findComponent('cancelButton').props('size')).toBe('small');
    });

    it('renders the correct cancel button', () => {
      expect(findComponent('cancelButton').props('type')).toBe('tertiary');
      expect(findComponent('cancelButton').props('size')).toBe('small');
      expect(findComponent('cancelButton').props('text')).toBe(
        i18n.global.t('agent_builder.instructions.edit_instruction.cancel'),
      );
    });

    describe('Save button', () => {
      it('is disabled when the input is empty', async () => {
        wrapper.vm.editingText = '';
        await wrapper.vm.$nextTick();
        expect(findComponent('saveButton').props('disabled')).toBe(true);
      });

      it('is enabled when the input is not empty', async () => {
        expect(findComponent('saveButton').props('disabled')).toBe(false);
      });

      it('is loading when the instruction is being edited', async () => {
        expect(findComponent('saveButton').props('loading')).toBe(false);
        wrapper.setProps({
          instruction: { ...wrapper.props.instruction, status: 'loading' },
        });
        await wrapper.vm.$nextTick();
        expect(findComponent('saveButton').props('loading')).toBe(true);
      });

      it('toggles edit mode when clicked', async () => {
        findComponent('saveButton').vm.$emit('click');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.isEditing).toBe(false);
      });
    });

    describe('Cancel button', () => {
      it('toggles edit mode when clicked', async () => {
        findComponent('cancelButton').vm.$emit('click');
        await wrapper.vm.$nextTick();
        expect(wrapper.vm.isEditing).toBe(false);
      });
    });
  });
});
