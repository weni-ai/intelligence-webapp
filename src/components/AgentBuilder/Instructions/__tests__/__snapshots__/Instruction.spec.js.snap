// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Instruction.vue > Component rendering > should match snapshot 1`] = `
"<section data-v-cba3a9e7="" class="instruction">
  <p data-v-cba3a9e7="" class="instruction__text" data-testid="instruction-text">Test instruction</p>
  <!--v-if-->
  <content-item-actions-stub data-v-cba3a9e7="" actions="[object Object],[object Object]" triggericon="more_vert" triggersize="avatar-nano" popoverpositionhorizontal="right-right" popoverpositionvertical="top-bottom" minwidth="170px" data-testid="instruction-actions"></content-item-actions-stub>
</section>
<modal-remove-instruction-stub data-v-cba3a9e7="" instruction="[object Object]" modelvalue="false" data-testid="modal-remove-instruction"></modal-remove-instruction-stub>"
`;

exports[`Instruction.vue > Edit mode > should match snapshot > edit-mode 1`] = `
"<section data-v-cba3a9e7="" class="instruction">
  <section data-v-cba3a9e7="" class="instruction__input">
    <unnnic-input-stub data-v-cba3a9e7="" placeholder="" type="normal" modelvalue="Test instruction" nativetype="text" allowtogglepassword="false" hascloudycolor="false" size="sm" mask="" class="input__field" data-testid="instruction-input"></unnnic-input-stub>
  </section>
  <unnnic-button-stub data-v-cba3a9e7="" size="small" text="Save" type="secondary" float="false" iconleft="" iconright="" iconcenter="" iconsfilled="false" next="false" disabled="false" loading="false" data-testid="instruction-save-button"></unnnic-button-stub>
  <unnnic-button-stub data-v-cba3a9e7="" size="small" text="Cancel" type="tertiary" float="false" iconleft="" iconright="" iconcenter="" iconsfilled="false" next="false" disabled="false" loading="false" data-testid="instruction-cancel-button"></unnnic-button-stub>
</section>
<modal-remove-instruction-stub data-v-cba3a9e7="" instruction="[object Object]" modelvalue="false" data-testid="modal-remove-instruction"></modal-remove-instruction-stub>"
`;
